applications/MappingApplication/custom_utilities/projection_utilities.cpp

```diff
@@ -0,0 +1,273 @@
+                               std::vector<int>& rEquationIds,
+                               double& rProjectionDistance,
+                               const bool ComputeApproximation)
+{
+    array_1d<double, 3> local_coords;
+    PairingIndex pairing_index;
+    bool is_inside = rGeometry.IsInside(rPointToProject, local_coords);
+
+    if (is_inside) {
+        pairing_index = PairingIndex::Volume_Inside;
+        rGeometry.ShapeFunctionsValues(rShapeFunctionValues, local_coords);
+        FillEquationIdVector(rGeometry, rEquationIds);
+
+        rProjectionDistance = MapperUtilities::ComputeDistance(rPointToProject, rGeometry.Center());
+        rProjectionDistance /= rGeometry.Volume(); // Normalize Distance by Volume
+
+    } else if (!ComputeApproximation) {
+        return PairingIndex::Unspecified;
+
+    } else if (!is_inside && rGeometry.IsInside(rPointToProject, local_coords, LocalCoordTol)) {
+        pairing_index = PairingIndex::Volume_Outside;
+        rGeometry.ShapeFunctionsValues(rShapeFunctionValues, local_coords);
+        FillEquationIdVector(rGeometry, rEquationIds);
+
+        rProjectionDistance = MapperUtilities::ComputeDistance(rPointToProject, rGeometry.Center());
+        rProjectionDistance /= rGeometry.Volume(); // Normalize Distance by Volume
+
+    } else { // inter-/extrapolation failed, trying to project on "subgeometries"
+        pairing_index = PairingIndex::Unspecified;
+        std::vector<int> face_eq_ids;
+        Vector face_sf_values;
+        double face_distance;
+
+        for (const auto& r_face : const_cast<GeometryType&>(rGeometry).Faces()) { // TODO update this after #4767
+
+            const PairingIndex face_index = ProjectOnSurface(r_face, rPointToProject, LocalCoordTol, face_sf_values, face_eq_ids, face_distance);
+
+            // check if the current edge gives a better result
+            if (face_index > pairing_index || (face_index == pairing_index && face_distance < rProjectionDistance)) {
+                pairing_index = face_index;
+                rShapeFunctionValues = face_sf_values;
+                rProjectionDistance = face_distance;
+                rEquationIds = face_eq_ids;
+            }
+        }
+    }
+
+    return pairing_index;
+}
+
+bool ComputeProjection(const GeometryType& rGeometry,
+                       const Point& rPointToProject,
+                       const double LocalCoordTol,
+                       Vector& rShapeFunctionValues,
+                       std::vector<int>& rEquationIds,
+                       double& rProjectionDistance,
+                       PairingIndex& rPairingIndex,
+                       const bool ComputeApproximation)
+{
+    const SizeType num_points = rGeometry.PointsNumber();
+    const auto geom_family = rGeometry.GetGeometryFamily();
+    bool is_full_projection = false;
+
+    if (geom_family == GeometryData::Kratos_Linear && num_points == 2) { // linear line
```

Comment:
You may want to check `KratosGeometryType` instead (would avoid the second condition)

Reply:
actually it would not, since then I have to check for `Kratos_Line2D2` or `Kratos_Line3D2`
Hence I would prefer to leave it as is